### DISCUSSION TITLE
make an example about how to watch key with certificate

### DISCUSSION THREAD

**Discussion by ruguoxiansheng** [etcd-io/jetcd#876] [Q&A]
Dear editor:
        Can you make an example about how to watch key with certificate with etcd?  like etcd-peer.pem,ca.pem. Thank you very much.

**Comment by lburgazzoli** [comment]
To configure SSL, jetcd relies entirely on what gRPC provides :

```java
 Client client = Client.builder()
    .endpoints(endpoint)
    .sslContext(builder -> {
        // do something with SslContextBuilder
    })
    .build();
```

**Reply by ruguoxiansheng** [reply]
Thank you very much. I have ca.pem,ca-key.pem,etcd.pem,etcd-peer.pem. But I am confused. Can you make detail?

**Reply by lburgazzoli** [reply]
you may have a look at this example: https://github.com/etcd-io/jetcd/blob/master/jetcd-core/src/test/java/io/etcd/jetcd/SslTest.java but beside that, you need to investigate how to use Netty's SslContextBuilder. I'm sorry but I don't have much time these days to work on an example

**Reply by ruguoxiansheng** [reply]
Dear Iburgazzoli:
      Happy Spring Festival！  I have take an example by jetcd-all-0.5.4.jar from maven , i failed.The error is like 
`java.lang.ClassNotFoundException: io.etcd.jetcd.shaded.io.netty.internal.tcnative.SSLContext` 
`java.lang.ClassNotFoundException: org.conscrypt.Conscrypt`
`java.lang.ClassNotFoundException: org/eclipse/jetty/alpn/ALPN`
`java.lang.IllegalStateException: Could not find TLS ALPN provider; no working netty-tcnative, Conscrypt, or Jetty NPN/ALPN available`
when i find the source from jar,i can't find the class file 'io.etcd.jetcd.shaded.io.netty.internal.tcnative.SSLContext' and package 'io.etcd.jetcd.shaded.io.netty.internal.tcnative';
and the error location in the GrpcSslContexts.java 
`
 private static SslProvider defaultSslProvider() {
        if (OpenSsl.isAvailable()) {
            logger.log(Level.FINE, "Selecting OPENSSL");
            return SslProvider.OPENSSL;
        } else {
            Provider provider = findJdkProvider();
            if (provider != null) {
                logger.log(Level.FINE, "Selecting JDK with provider {0}", provider);
                return SslProvider.JDK;
            } else {
                logger.log(Level.INFO, "Java 9 ALPN API unavailable (this may be normal)");
                logger.log(Level.INFO, "netty-tcnative unavailable (this may be normal)", OpenSsl.unavailabilityCause());
                logger.log(Level.INFO, "Conscrypt not found (this may be normal)", GrpcSslContexts.ConscryptHolder.UNAVAILABILITY_CAUSE);
                logger.log(Level.INFO, "Jetty ALPN unavailable (this may be normal)", JettyTlsUtil.getJettyAlpnUnavailabilityCause());
                throw new IllegalStateException("Could not find TLS ALPN provider; no working netty-tcnative, Conscrypt, or Jetty NPN/ALPN available");
            }
        }
    }
`


My codes like this,and the etcd server use GO language.
`
public static synchronized Client getEtclClient()
    {
        if (etcdClient == null)
        {
            //String[] urls = PropertiesUtil.getValue("etcd_node_url").split(",");
            SslContext sslContext = null;

            try {
                final InputStream is = new FileInputStream(new File("D:\\Users\\jetwu\\Desktop\\ca.pem"));
                final InputStream certFile = new FileInputStream(new File("D:\\Users\\jetwu\\Desktop\\etcd-peer.pem"));
                final InputStream keyFile = new FileInputStream(new File("D:\\Users\\jetwu\\Desktop\\etcd-peer-key.pem"));
                String[] urls = "http://127.0.0.1:2379".split(",");
                SslContext builder = GrpcSslContexts.forClient().trustManager(is).keyManager(certFile,keyFile).build();
                etcdClient = Client.builder().endpoints(urls).authority("etcd0").
                        sslContext(builder).build();

            } catch (FileNotFoundException e) {
                e.printStackTrace();
            } catch (SSLException e) {
                e.printStackTrace();
            }
//

        }
        return etcdClient;
    }

`

can you help me?Thank you sincerely！

**Reply by lburgazzoli** [reply]
can you please try to use individual jars instead of jetcd-all ?
